### PR TITLE
fix: wrong order of messages from archive

### LINKED ISSUE #999: Old messages load failure with specific message
Hi team, I have an issue regarding the load of old messages.
I've found that if you send a message with some special characters like "$ % & ( ) * / \ + - < >" there's no problem. But when you try to load old messages, and within these messages there's one like that, jsxc seems to enter in a loop and stop working (console logs -> console.trace() Loop detected 8ab70b02-3f6f-4e57-a876-894f2351af4b).

I've found that in the browser localStorage the json linked to that message ("$ % & ( ) * / \ + - < >") seems to have the same value on "attrId" and "next": here's an example

`jsxc2:8ab70b02-3f6f-4e57-a876-894f2351af4b:"{"unread":false,"mark":2,"encrypted":null,"forwarded":false,"stamp":1609345781035,"type":"chat","encryptedHtmlMessage":null,"encryptedPlaintextMessage":null,"attrId":"8ab70b02-3f6f-4e57-a876-894f2351af4b","peer":"michael2@mydomain","direction":1,"plaintextMessage":"$ % & ( ) * / \\ + - < >","next":"8ab70b02-3f6f-4e57-a876-894f2351af4b"}"`

This happens ONLY when that message is in the message history. Instead, on the moment when you try to send the "$ % & ( ) * / \\ + - < >" message,  this is correct on the localStorage (I mean, the values on "attrId" and "next" inside its json are different).

### Expected behavior
Any message is recovered correctly from the message history

### Actual behavior
There's a case of one specific message containing "$ % & ( ) * / \\ + - < >" that fires an infinte loop.

### Steps to reproduce the behavior
1. Send a message "$ % & ( ) * / \\ + - < >"
2. Make it "old" sending other messages
3. Try to recover the old messages

### Environment
- JSXC version: 4.1.1
- JSXC is integrated on a proprietary web app
- Firefox v 84.0.2 (and probably Edge and Chrome, latest versions)
- No plugins enabled  on the browser
- XMPP Server: Openfire v 4.5.2 and 4.6.0 
- Tried to reproduce the behaviour on demo page, but the "load old messages" button seems not working there

#### JSXC
console.trace() Loop detected 8ab70b02-3f6f-4e57-a876-894f2351af4b

#### XMPP
No relevant messages on server logs

### LINKED ISSUE #1002: Catch Wrong/Reversed order when retrieving messages from archive...
### Expected behavior
When using "Load old messages" the ordering of messages should be the same as it was when the messages were written before...

Message order should be (top to bottom) A, B, C, D, E, F
### Actual behavior
When i try to retrieve message from archive via xep313 ("Load old messages" Button) the oldest message is on bottom and the newest on top.
Message order is (top to bottom) F, E, D, C, B, A

### Steps to reproduce the behavior
1. login
2. open chatbox
3. click on "load old messages" button

### Environment
- **JSXC version:** actual master, but same problem on your demo app
- **Host system and version:** ...
- **Browser vendor and version:** Firefox and Chrome
- **Any browser plugins enabled?**  no
- **XMPP server vendor and version:** Openfire 4.6 & Message Archivemenagement Plugin 2.2.0
- **Is your XMPP server working with other clients as expected?** Yes Converse and Gajim show messages in right order

#### Javascript

```
no error messages or messages related to mam other than the following in xmpp section
```

#### JSXC

```
no error messages or messages related to mam other than the following in xmpp section
```

#### XMPP

This comes from the server log
```
<iq id="b74c8082-54a4-4770-b5ef-fa5658665f10:sendIQ" to="mops@mydomain.local" type="set" from="mops@mydomain.local/jsxc-b69642ac"><query xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019"><x xmlns="jabber:x:data" type="submit"><field type="hidden" var="FORM_TYPE"><value>urn:xmpp:mam:2</value></field><field var="with"><value>admin@mydomain.local</value></field></x><set xmlns="http://jabber.org/protocol/rsm"><max>20</max><before/></set></query></iq>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3263"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:57.372Z"/><message xmlns="jabber:client" id="131f9520-ac13-4cff-b65e-baa8d28c7b90" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>F</body><origin-id xmlns="urn:xmpp:sid:0" id="131f9520-ac13-4cff-b65e-baa8d28c7b90"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3262"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:55.759Z"/><message xmlns="jabber:client" id="97b11150-d4d8-4ed3-b5c0-955954d3f91f" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>E</body><origin-id xmlns="urn:xmpp:sid:0" id="97b11150-d4d8-4ed3-b5c0-955954d3f91f"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3261"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:54.186Z"/><message xmlns="jabber:client" id="b133b07c-9c8e-4320-99a2-fadff47dd281" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>D</body><origin-id xmlns="urn:xmpp:sid:0" id="b133b07c-9c8e-4320-99a2-fadff47dd281"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3260"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:52.670Z"/><message xmlns="jabber:client" id="49b15b84-b6de-4285-b100-01bff702f0ab" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>C</body><origin-id xmlns="urn:xmpp:sid:0" id="49b15b84-b6de-4285-b100-01bff702f0ab"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3259"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:51.217Z"/><message xmlns="jabber:client" id="2e9ba264-4f01-4de8-830f-f676b33c7746" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>B</body><origin-id xmlns="urn:xmpp:sid:0" id="2e9ba264-4f01-4de8-830f-f676b33c7746"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<message to="mops@mydomain.local/jsxc-b69642ac"><result xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019" id="3258"><forwarded xmlns="urn:xmpp:forward:0"><delay xmlns="urn:xmpp:delay" stamp="2021-01-13T20:11:50.371Z"/><message xmlns="jabber:client" id="7ac40b3d-d209-461f-ac02-341414e9d52c" to="admin@mydomain.local" type="chat" from="mops@mydomain.local/jsxc-99752c3c"><body>A</body><origin-id xmlns="urn:xmpp:sid:0" id="7ac40b3d-d209-461f-ac02-341414e9d52c"></origin-id><request xmlns="urn:xmpp:receipts"></request><active xmlns="http://jabber.org/protocol/chatstates"></active><markable xmlns="urn:xmpp:chat-markers:0"></markable></message></forwarded></result></message>

<iq type="result" id="b74c8082-54a4-4770-b5ef-fa5658665f10:sendIQ" from="mops@mydomain.local" to="mops@mydomain.local/jsxc-b69642ac"><fin xmlns="urn:xmpp:mam:2" queryid="e01b5704-e2dd-488b-a598-2579c81a1019"><set xmlns="http://jabber.org/protocol/rsm"><first>3263</first><last>3258</last><count>6</count></set></fin></iq>
```

i noticed the id (3263) in result iq the in "&lt;first&gt;" is higher than the id (3258) in "&lt;last&gt;"
&lt;set xmlns="http://jabber.org/protocol/rsm" &gt; &lt;first&gt;3263&lt;/first&gt;&lt;last&gt;3258&lt;/last&gt;&lt;count&gt;6&lt;/count&gt;&lt;/set&gt;...

May be its related to the servers implementation and depends on database query result ordering (asc or desc)... but this could be catched if we compare the &lt;first&gt; tag with &lt;last&gt; tag and do reverse the messages if &lt;first&gt; is higher than &lt;last&gt;

Further more because of the wrong order every further click on "load old messages" will result in a corrupted message order in the chatbox then...

maybe this issue https://github.com/jsxc/jsxc/issues/999 is related to this here too...

### PR DESCRIPTION
**Pull request by mightymop**
fix #1002
seems to fix #999 too on my quick testings

this behavoir relies on: https://github.com/igniterealtime/openfire-monitoring-plugin/issues/165

### CODE REVIEW

**Comment by mightymop** [comment]
I know you are not a fan of OF, but nevertheless this case should be catched.

File: src/plugins/mam/Archive.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ import MultiUserContact from '@src/MultiUserContact';
 
 export default class Archive {
    private archiveJid: IJID;
-   private messageCache: JQuery<HTMLElement>[] = [];
+   private messageCache: any[] = [];
```
Comment (sualko) [inline src/plugins/mam/Archive.ts:16]: That's never a good idea. Seams like a source for massive issues in the future.

File: src/plugins/mam/Archive.ts [hunk 2]
```diff
@@ -96,8 +96,9 @@ export default class Archive {
          });
    }
 
-   public onForwardedMessage(forwardedElement: JQuery<HTMLElement>) {
-      this.messageCache.push(forwardedElement);
+   public onForwardedMessage(forwardedElement: JQuery<HTMLElement>, messageid : string) {
```
Comment (sualko) [inline src/plugins/mam/Archive.ts:99]: That's the result id, right?

File: src/plugins/mam/Archive.ts [hunk 3]
```diff
@@ -172,8 +173,29 @@ export default class Archive {
 
       let transcript = this.contact.getTranscript();
 
+      let isArchiveExhausted = finElement.attr('complete') === 'true';
+      let queryId = finElement.attr('queryid');
```
Comment (sualko) [inline src/plugins/mam/Archive.ts:177]: Why was this moved up?

File: src/plugins/mam/Plugin.ts [hunk 4]
```diff
@@ -174,6 +174,7 @@ export default class MessageArchiveManagementPlugin extends AbstractPlugin {
    private onMamMessage = (stanza: string): boolean => {
       let stanzaElement = $(stanza);
       let resultElement = stanzaElement.find(`result[xmlns^="urn:xmpp:mam:"]`);
+      let id = resultElement.attr('id');
```
Comment (sualko) [inline src/plugins/mam/Plugin.ts:177]: There are many ids, so they should be named properly.

**Review by sualko — CHANGES_REQUESTED** [review]
Only a few Typescript enhancements.

File: src/plugins/mam/Plugin.ts (hunk 4 above)
Comment (mightymop) [reply src/plugins/mam/Plugin.ts]: renamed

File: src/plugins/mam/Archive.ts (hunk 3 above)
Comment (mightymop) [reply src/plugins/mam/Archive.ts:183]: i thought it would make the code more readable if the vars are set before the loops and if clauses...

File: src/plugins/mam/Archive.ts (hunk 2 above)
Comment (mightymop) [reply src/plugins/mam/Archive.ts]: right.. i renamed it

File: src/plugins/mam/Archive.ts (hunk 1 above)
Comment (mightymop) [reply src/plugins/mam/Archive.ts]: i created a simple class for that now

**Comment by mightymop** [comment]
The bug in OF MAM plugin is now fixed in latest Release 2.2.1. So now this is a minor improval if we would add this catch

**Comment by sualko** [comment]
Since the issue is fixed in OF, I close this pr. Thanks for all the effort you put on improving XMPP :+1:
